README.md
# Flipt Validate Action

![GitHub Release](https://img.shields.io/github/v/release/flipt-io/validate-action)

This action validates [Flipt](https://flipt.io) feature flag [features.yaml](https://www.flipt.io/docs/experimental/filesystem-backends#defining-flag-state) files for syntax and semantic errors.

![Validate](./.github/images/action.png)

## Example

```yaml
# testing/features.yaml
namespace: default
flags:
  - key: flipt
    name: flipt
    description: flipt
    enabled: false
    variants:
      - key: flipt
        name: flipt
      - key: flipt
        name: flipt
    rules:
      - segment: internal-users
        rank: 1
        distributions:
          - variant: fromFlipt
            rollout: 110
      - segment: all-users
        rank: 2
        distributions:
          - variant: fromFlipt2
            rollout: 100
```

Would result in the following output:

```console
Validation failed!

- Message  : flags.0.rules.0.distributions.0.rollout: invalid value 110 (out of bound <=100)
  File     : testing/features.yaml
  Line     : 17
  Column   : 23
```

## Usage

```yaml
validate:
  runs-on: ubuntu-latest
  steps:
    # Checkout the target repository
    - uses: actions/checkout@v3

    - uses: flipt-io/validate-action@v0.1.0
      # with:
        # Optional, the token to use for GitHub API requests
        # github-token: ${{ secrets.GITHUB_TOKEN }}
        # Optional, the directory to validate, defaults to the repository root
        # working-directory:
```

## Customizing

### inputs

Following inputs can be used as `step.with` keys

| Name                | Type   | Description                                                                |
| ------------------- | ------ | -------------------------------------------------------------------------- |
| `github-token`      | string | **Optional**. The token to use for GitHub API requests                     |
| `working-directory` | string | **Optional**. The directory to validate, defaults to the repository root   |
| `args`              | string | **Optional**. Additional arguments to pass to the `flipt validate` command |

## Development

```console
# Builds the typescript code.
npm run build

# Runs eslint.
npm run lint

# Runs prettier.
npm run format

# Packages the code into the dist folder. Must be updated to pass CI on `main`.
npm run package

# Runs all of the above commands.
npm run all
```

## License

[Apache 2.0](LICENSE)


CHANGELOG.md
# Changelog

## [0.2.0](https://github.com/flipt-io/validate-action/compare/v0.1.0...v0.2.0) (2023-12-11)


### Bug Fixes

* validate no longer takes files ([#88](https://github.com/flipt-io/validate-action/issues/88)) ([1851d1c](https://github.com/flipt-io/validate-action/commit/1851d1c6288012b8069435de0db076145af96f00))


### Miscellaneous Chores

* release 0.2.0 ([10c60e8](https://github.com/flipt-io/validate-action/commit/10c60e87e4303064951f4ec42ea0f9b5af1aefd1))

## 0.1.0 (2023-08-08)


### Features

* add validate impl; integration test ([dd718a3](https://github.com/flipt-io/validate-action/commit/dd718a358601f0b3cf3439c1d00606ade45856c7))
* make GITHUB_TOKEN optional for flipt dl ([40ddf95](https://github.com/flipt-io/validate-action/commit/40ddf959cf285fc1a4c1c37c1f855afa5408a61e))
* remove docker requirement ([4c22b6c](https://github.com/flipt-io/validate-action/commit/4c22b6cef03ace159c3b7b8027aabb60125e680d))


### Bug Fixes

* dont pass args ([2590c98](https://github.com/flipt-io/validate-action/commit/2590c9824c39fe4cdcc5271a780cdbfe77a996ff))
* release-please action ([c5f1b6a](https://github.com/flipt-io/validate-action/commit/c5f1b6aee543245d2aedb6781fd02c656a7d3e26))
* rm need for github token, specify flipt version ([b78c519](https://github.com/flipt-io/validate-action/commit/b78c51979656c95c0efd175633e9d5a14c48ee43))
* use correct workspace path ([ca81780](https://github.com/flipt-io/validate-action/commit/ca81780091ccac5388fc55b14e70dffeab8a996f))
* version arg ([ad588f1](https://github.com/flipt-io/validate-action/commit/ad588f15d859ed6884b6596d847691909ce90cb4))


### Miscellaneous Chores

* release 0.1.0 ([2a3c906](https://github.com/flipt-io/validate-action/commit/2a3c906639a40a3c25aeb5193bd51f690c973184))


src/lib/cli.ts
import * as os from 'os';
import * as core from '@actions/core';
import * as io from '@actions/io';
import * as github from '@actions/github';
import * as tc from '@actions/tool-cache';
import { exec } from './exec';
import path from 'path';
import fetch from "node-fetch";
import { Octokit } from '@octokit/rest';

function getPlatform(): string | undefined {
  const platforms = {
    'linux-x64': 'linux_x86_64',
    // 'linux-arm64': 'linux_arm64',
    // 'darwin-x64': 'darwin-x64',
    // 'darwin-arm64': 'darwin-arm64',
    // 'win32-x64': 'windows-x64'
  };

  const runnerPlatform = os.platform();
  const runnerArch = os.arch();

  return platforms[`${runnerPlatform}-${runnerArch}` as keyof typeof platforms];
}

export async function downloadFlipt(): Promise<void> {
  let octokit: any;

  const repoToken: string = core.getInput('github-token', { required: false })
  if (repoToken) {
    core.debug('Using provided github token');
    octokit = github.getOctokit(repoToken);
  } else {
    core.debug('No github token provided');
    octokit = new Octokit({
      request: {fetch: fetch},
    });
  }

  const platform = getPlatform();
  if (!platform) {
    throw new Error('Unsupported platform');
  }

  // check if flipt is already installed
  // if not, download it
  const isFliptInstalled = await io.which('flipt');

  if (isFliptInstalled) {
    core.info('flipt is already installed, skipping download');
    return;
  }

  core.info('flipt is not installed, proceeding to downloading');

  try {
    const release = await octokit.rest.repos.getLatestRelease({
      owner: 'flipt-io',
      repo: 'flipt',
    });

    const version = release.data.tag_name;

    core.debug(`Latest flipt release is ${version}`);

    const asset = release.data.assets.find(
      (asset: { name: string; }) => asset.name === `flipt_${platform}.tar.gz`,
    );

    const downloadUrl = asset?.browser_download_url;

    if (!downloadUrl) {
      throw new Error(`No download url found for ${platform}`);
    }

    core.debug(`Downloading from ${downloadUrl}`);

    const destination = path.join(os.homedir(), '.flipt/flipt');
    core.debug(`Install destination is ${destination}`);

    await io
      .rmRF(destination)
      .catch()
      .then(() => {
        core.debug(
          `Successfully deleted pre-existing ${destination} directory (if any)}`,
        );
      });

    const downloaded = await tc.downloadTool(downloadUrl);

    core.debug(`Successfully downloaded 'flipt-${version}' to ${downloaded}`);

    await io.mkdirP(destination);
    const extractedPath = await tc.extractTar(downloaded, destination);
    core.debug(`Successfully extracted ${downloaded} to ${extractedPath}`);

    const cachedPath = await tc.cacheDir(destination, 'flipt', version);

    core.addPath(cachedPath);
    core.debug(`Successfully cached ${destination} to ${cachedPath}`);

    const versionExec = await exec('flipt', ['--version'], true);
    if (!versionExec.success) {
      throw new Error(`flipt failed to run: ${versionExec.stderr.trim()}`);
    }

    core.info(`${versionExec.stdout.trim()} installed`);
  } catch (error) {
    throw new Error(`Failed to fetch latest release: ${error}`);
  }
}


src/main.ts
import * as core from '@actions/core'
import {downloadFlipt} from './lib/cli'
import {environmentVariables} from './lib/environment'
import {exec} from './lib/exec'

async function run(): Promise<void> {
  try {
    const argsInput = core.getInput('args')

    // split args by whitespace
    const args = argsInput
      .trim()
      .split(/\s+/)
      .filter(arg => arg !== '')
      .filter(arg => arg.length > 0)

    await validate(args)
  } catch (error) {
    if (error instanceof Error) core.setFailed(error.message)
  }
}

async function validate(args: string[] = []): Promise<void> {
  core.startGroup(`Installing flipt:latest`)
  await downloadFlipt()
  core.endGroup()

  let workspace = core.getInput('working-directory')

  if (!workspace || workspace.length === 0) {
    workspace = environmentVariables.GITHUB_WORKSPACE
  }

  const continueOnError = args.includes('--continue-on-error')

  core.debug(`running flipt in workspace: ${workspace}`)

  core.startGroup('Running flipt validate')

  const result = await exec(
    'flipt',
    ['validate', '--issue-exit-code=0', '--format=json'],
    {
      env: {
        FLIPT_LOG_LEVEL: 'fatal'
      }
    }
  )

  core.endGroup()

  if (!result.success) {
    core.setFailed(`flipt validate failed: ${result.stderr}`)
    return
  }

  // TODO: hack to get second line of output until https://github.com/flipt-io/flipt/issues/2531 is fixed
  const lines = result.stdout.toString().split('\n')
  if (lines.length < 2) {
    core.setFailed('flipt validate returned invalid output')
    return
  }

  // get second line of output
  const line = lines[1]

  const response = line.trim()

  core.debug(`flipt response: ${response}`)

  if (response.length === 0) {
    core.debug('flipt returned empty response')
    core.info('✅ No invalid files found')
    return
  }

  let errors = 0
  const json = JSON.parse(response)

  // response looks like:
  // [{"message":"flags.0.rules.0.distributions.0.rollout: invalid value 110 (out of bound \\u003c=100)","location":{"file":"features.yaml","line":17}}]

  // loop through objects in response
  for (const error of json) {
    errors += 1
    const {message, location} = error

    // TODO: hack
    const file = location.file.replace('/workspace/', '')

    const opts: core.AnnotationProperties = {
      file,
      startLine: location.line
    }

    core.error(message, opts)
  }

  const msg = `Found ${errors} error(s) when validating features.yml`
  if (continueOnError) {
    core.warning(msg)
    return
  }
  core.setFailed(msg)
}

run()
